<!DOCTYPE html>
<html>
<head><title>R: Model-based Analysis of Single Cell Transcriptomics</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> Model-based Analysis of Single Cell Transcriptomics
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;MAST&rsquo; version 1.22.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<p style="text-align: center;">
<a href="# "> </a>
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#H">H</a>
<a href="#I">I</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#U">U</a>
<a href="#V">V</a>
<a href="#W">W</a>
<a href="#X">X</a>
<a href="#Z">Z</a>
<a href="#misc">misc</a>
</p>

<table width="100%">
<tr><td style="width: 25%;"><a href="MAST-package.html">MAST-package</a></td>
<td>MAST: Model-based Analysis of Single- cell Transcriptomics</td></tr>
</table>

<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="applyFlat.html">applyFlat</a></td>
<td>Apply a vectorized binary operation recycling over last dimension</td></tr>
<tr><td style="width: 25%;"><a href="defaultAssay.html">assay-method</a></td>
<td>Default 'assay' returned</td></tr>
<tr><td style="width: 25%;"><a href="defaultAssay.html">assay_idx</a></td>
<td>Default 'assay' returned</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="BayesGLMlike-class.html">BayesGLMlike-class</a></td>
<td>Wrapper for bayesian GLM</td></tr>
<tr><td style="width: 25%;"><a href="bootVcov1.html">bootVcov1</a></td>
<td>Bootstrap a zlmfit</td></tr>
<tr><td style="width: 25%;"><a href="mast_filter.html">burdenOfFiltering</a></td>
<td>Filter a SingleCellAssay</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="calcZ.html">calcZ</a></td>
<td>Get Z or T statistics and P values after running gseaAfterBoot</td></tr>
<tr><td style="width: 25%;"><a href="MAST-defunct.html">cData</a></td>
<td>Defunct functions in package 'MAST'</td></tr>
<tr><td style="width: 25%;"><a href="LMERlike-class.html">coef-method</a></td>
<td>Wrapper for lmer/glmer</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">coef-method</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
<tr><td style="width: 25%;"><a href="Hypothesis.html">CoefficientHypothesis</a></td>
<td>Describe a linear model hypothesis to be tested</td></tr>
<tr><td style="width: 25%;"><a href="colData-set-SingleCellAssay-DataFrame-method.html">colData&lt;--method</a></td>
<td>Replace 'colData'</td></tr>
<tr><td style="width: 25%;"><a href="collectResiduals.html">collectResiduals</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="collectResiduals.html">combined_residuals_hook</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="computeEtFromCt.html">computeEtFromCt</a></td>
<td>Compute the Et from the Ct</td></tr>
<tr><td style="width: 25%;"><a href="freq.html">condmean</a></td>
<td>Summary statistics for genes in an experiment</td></tr>
<tr><td style="width: 25%;"><a href="freq.html">condSd</a></td>
<td>Summary statistics for genes in an experiment</td></tr>
<tr><td style="width: 25%;"><a href="collectResiduals.html">continuous_residuals_hook</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="convertMASTClassicToSingleCellAssay.html">convertMASTClassicToSingleCellAssay</a></td>
<td>Convert a MASTClassic SingleCellAssay</td></tr>
<tr><td style="width: 25%;"><a href="CovFromBoots.html">CovFromBoots</a></td>
<td>Extract the inter-gene covariance matrices for continuous and discrete components of a MAST model for a given coefficient from bootstrap replicates</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="defaultAssay.html">defaultAssay</a></td>
<td>Default 'assay' returned</td></tr>
<tr><td style="width: 25%;"><a href="defaultPrior.html">defaultPrior</a></td>
<td>Initialize a prior to be used a prior for BayeGLMlike/BayesGLMlike2</td></tr>
<tr><td style="width: 25%;"><a href="collectResiduals.html">deviance_residuals_hook</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="collectResiduals.html">discrete_residuals_hook</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="dof.html">dof</a></td>
<td>Degrees of freedom of Zero inflated model</td></tr>
<tr><td style="width: 25%;"><a href="dof.html">dof-method</a></td>
<td>Degrees of freedom of Zero inflated model</td></tr>
<tr><td style="width: 25%;"><a href="Drop.html">Drop</a></td>
<td>Drop specified dimension from an array</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="ebayes.html">ebayes</a></td>
<td>Estimate hyperparameters for hierarchical variance model for continuous component</td></tr>
<tr><td style="width: 25%;"><a href="expavg.html">expavg</a></td>
<td>Exponential average</td></tr>
<tr><td style="width: 25%;"><a href="MAST-defunct.html">exprs</a></td>
<td>Defunct functions in package 'MAST'</td></tr>
</table>

<h2><a name="F">-- F --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="MAST-defunct.html">fData</a></td>
<td>Defunct functions in package 'MAST'</td></tr>
<tr><td style="width: 25%;"><a href="MAST-defunct.html">filter</a></td>
<td>Defunct functions in package 'MAST'</td></tr>
<tr><td style="width: 25%;"><a href="filterLowExpressedGenes.html">filterLowExpressedGenes</a></td>
<td>Filter low-expressing genes</td></tr>
<tr><td style="width: 25%;"><a href="fit.html">fit</a></td>
<td>fit a zero-inflated regression</td></tr>
<tr><td style="width: 25%;"><a href="fit.html">fit-method</a></td>
<td>fit a zero-inflated regression</td></tr>
<tr><td style="width: 25%;"><a href="collectResiduals.html">fitted_phat</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="FromFlatDF.html">FluidigmAssay</a></td>
<td>Construct a SingleCellAssay (or derived subclass) from a 'flat' (melted) data.frame/data.table</td></tr>
<tr><td style="width: 25%;"><a href="freq.html">freq</a></td>
<td>Summary statistics for genes in an experiment</td></tr>
<tr><td style="width: 25%;"><a href="FromFlatDF.html">FromFlatDF</a></td>
<td>Construct a SingleCellAssay (or derived subclass) from a 'flat' (melted) data.frame/data.table</td></tr>
<tr><td style="width: 25%;"><a href="FromMatrix.html">FromMatrix</a></td>
<td>Construct a SingleCellAssay from a matrix or array of expression</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="getConcordance.html">getConcordance</a></td>
<td>Get the concordance between two experiments</td></tr>
<tr><td style="width: 25%;"><a href="logFC.html">getLogFC</a></td>
<td>Calculate log-fold changes from hurdle model components</td></tr>
<tr><td style="width: 25%;"><a href="getConcordance.html">getrc</a></td>
<td>Get the concordance between two experiments</td></tr>
<tr><td style="width: 25%;"><a href="getConcordance.html">getss</a></td>
<td>Get the concordance between two experiments</td></tr>
<tr><td style="width: 25%;"><a href="getwellKey.html">getwellKey</a></td>
<td>Accessor for wellKey</td></tr>
<tr><td style="width: 25%;"><a href="getwellKey.html">getwellKey-method</a></td>
<td>Accessor for wellKey</td></tr>
<tr><td style="width: 25%;"><a href="getConcordance.html">getwss</a></td>
<td>Get the concordance between two experiments</td></tr>
<tr><td style="width: 25%;"><a href="GLMlike-class.html">GLMlike-class</a></td>
<td>Wrapper for regular glm/lm</td></tr>
<tr><td style="width: 25%;"><a href="gseaAfterBoot.html">gseaAfterBoot</a></td>
<td>Gene set analysis for hurdle model</td></tr>
<tr><td style="width: 25%;"><a href="GSEATests-class.html">GSEATests-class</a></td>
<td>An S4 class for Gene Set Enrichment output</td></tr>
<tr><td style="width: 25%;"><a href="gseaAfterBoot.html">gsea_control</a></td>
<td>Gene set analysis for hurdle model</td></tr>
</table>

<h2><a name="H">-- H --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="hushWarning.html">hushWarning</a></td>
<td>Selectively muffle warnings based on output</td></tr>
<tr><td style="width: 25%;"><a href="Hypothesis.html">Hypothesis</a></td>
<td>Describe a linear model hypothesis to be tested</td></tr>
</table>

<h2><a name="I">-- I --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="impute.html">impute</a></td>
<td>impute missing continuous expression for plotting</td></tr>
<tr><td style="width: 25%;"><a href="influence.bayesglm.html">influence.bayesglm</a></td>
<td>Influence bayesglm object</td></tr>
<tr><td style="width: 25%;"><a href="invlogit.html">invlogit</a></td>
<td>Inverse of logistic transformation</td></tr>
</table>

<h2><a name="L">-- L --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="LMERlike-class.html">LMERlike-class</a></td>
<td>Wrapper for lmer/glmer</td></tr>
<tr><td style="width: 25%;"><a href="LMlike-class.html">LMlike-class</a></td>
<td>Linear Model-like Class</td></tr>
<tr><td style="width: 25%;"><a href="logFC.html">logFC</a></td>
<td>Calculate log-fold changes from hurdle model components</td></tr>
<tr><td style="width: 25%;"><a href="LMERlike-class.html">logLik-method</a></td>
<td>Wrapper for lmer/glmer</td></tr>
<tr><td style="width: 25%;"><a href="LMlike-class.html">logLik-method</a></td>
<td>Linear Model-like Class</td></tr>
<tr><td style="width: 25%;"><a href="logmean.html">logmean</a></td>
<td>Log mean</td></tr>
<tr><td style="width: 25%;"><a href="LRT.html">LRT</a></td>
<td>Likelihood Ratio Tests for SingleCellAssays</td></tr>
<tr><td style="width: 25%;"><a href="LRT.html">LRT-method</a></td>
<td>Likelihood Ratio Tests for SingleCellAssays</td></tr>
<tr><td style="width: 25%;"><a href="lrTest.html">lrTest</a></td>
<td>Run a likelihood-ratio test</td></tr>
<tr><td style="width: 25%;"><a href="LMlike-class.html">lrTest-method</a></td>
<td>Linear Model-like Class</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">lrTest-method</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
<tr><td style="width: 25%;"><a href="lrTest-ZlmFit-character-method.html">lrTest-method</a></td>
<td>Likelihood ratio test</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="defaultAssay.html">magic_assay_names</a></td>
<td>Default 'assay' returned</td></tr>
<tr><td style="width: 25%;"><a href="maits-dataset.html">maits</a></td>
<td>MAITs data set, RNASeq</td></tr>
<tr><td style="width: 25%;"><a href="MAST-package.html">MAST</a></td>
<td>MAST: Model-based Analysis of Single- cell Transcriptomics</td></tr>
<tr><td style="width: 25%;"><a href="MAST-defunct.html">MAST-defunct</a></td>
<td>Defunct functions in package 'MAST'</td></tr>
<tr><td style="width: 25%;"><a href="mast_filter.html">mast_filter</a></td>
<td>Filter a SingleCellAssay</td></tr>
<tr><td style="width: 25%;"><a href="meld_list_left.html">meld_list_left</a></td>
<td>Combine lists, preferentially taking elements from x if there are duplicate names</td></tr>
<tr><td style="width: 25%;"><a href="melt.SingleCellAssay.html">melt.SingleCellAssay</a></td>
<td>"Melt" a 'SingleCellAssay' matrix</td></tr>
<tr><td style="width: 25%;"><a href="model.matrix.html">model.matrix</a></td>
<td>Model matrix accessor</td></tr>
<tr><td style="width: 25%;"><a href="model.matrix.html">model.matrix-method</a></td>
<td>Model matrix accessor</td></tr>
<tr><td style="width: 25%;"><a href="model.matrix-set.html">model.matrix&lt;-</a></td>
<td>Replace model matrix</td></tr>
<tr><td style="width: 25%;"><a href="myBiplot.html">myBiplot</a></td>
<td>Makes a nice BiPlot</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="freq.html">numexp</a></td>
<td>Summary statistics for genes in an experiment</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="collectResiduals.html">partialScore</a></td>
<td>Residual hooks and collection methods</td></tr>
<tr><td style="width: 25%;"><a href="bootVcov1.html">pbootVcov1</a></td>
<td>Bootstrap a zlmfit</td></tr>
<tr><td style="width: 25%;"><a href="plot.thresholdSCRNACountMatrix.html">plot.thresholdSCRNACountMatrix</a></td>
<td>Plot cutpoints and densities for thresholding</td></tr>
<tr><td style="width: 25%;"><a href="plotlrt.html">plotlrt</a></td>
<td>Plot a likelihood ratio test object</td></tr>
<tr><td style="width: 25%;"><a href="plotSCAConcordance.html">plotSCAConcordance</a></td>
<td>Concordance plots of filtered single vs n-cell assays</td></tr>
<tr><td style="width: 25%;"><a href="predict.ZlmFit.html">predict.ZlmFit</a></td>
<td>Return predictions from a ZlmFit object.</td></tr>
<tr><td style="width: 25%;"><a href="predicted_sig-dataset.html">predicted_sig</a></td>
<td>Predicted signatures</td></tr>
<tr><td style="width: 25%;"><a href="primerAverage.html">primerAverage</a></td>
<td>Average expression values for duplicated/redundant genes</td></tr>
<tr><td style="width: 25%;"><a href="summary.thresholdSCRNACountMatrix.html">print.summaryThresholdSCRNA</a></td>
<td>Summarize the effect of thresholding</td></tr>
<tr><td style="width: 25%;"><a href="print.summaryZlmFit.html">print.summaryZlmFit</a></td>
<td>Print summary of a ZlmFit</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="read.fluidigm.html">read.fluidigm</a></td>
<td>Reads a Fluidigm Biomark (c. 2011) raw data file (or set of files)</td></tr>
<tr><td style="width: 25%;"><a href="removeResponse.html">removeResponse</a></td>
<td>Remove the left hand side (response) from a formula</td></tr>
<tr><td style="width: 25%;"><a href="rstandard.bayesglm.html">rstandard.bayesglm</a></td>
<td>rstandard for bayesglm objects.</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="SceToSingleCellAssay.html">SceToSingleCellAssay</a></td>
<td>Coerce a SingleCellExperiment to some class defined in MAST</td></tr>
<tr><td style="width: 25%;"><a href="se.coef.html">se.coef</a></td>
<td>Return coefficient standard errors</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">se.coef-method</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
<tr><td style="width: 25%;"><a href="show.html">show-method</a></td>
<td>show</td></tr>
<tr><td style="width: 25%;"><a href="FromFlatDF.html">SingleCellAssay</a></td>
<td>Construct a SingleCellAssay (or derived subclass) from a 'flat' (melted) data.frame/data.table</td></tr>
<tr><td style="width: 25%;"><a href="split-SingleCellAssay-character-method.html">split-method</a></td>
<td>Split into 'list'</td></tr>
<tr><td style="width: 25%;"><a href="stat_ell.html">stat_ell</a></td>
<td>Plot confidence ellipse in 2D</td></tr>
<tr><td style="width: 25%;"><a href="subset-SingleCellAssay-method.html">subset-method</a></td>
<td>Subset a 'SingleCellAssay' by cells (columns)</td></tr>
<tr><td style="width: 25%;"><a href="summarize.html">summarize</a></td>
<td>Return programmatically useful summary of a fit</td></tr>
<tr><td style="width: 25%;"><a href="LMlike-class.html">summary-method</a></td>
<td>Linear Model-like Class</td></tr>
<tr><td style="width: 25%;"><a href="summary-GSEATests-method.html">summary-method</a></td>
<td>Summarize gene set enrichment tests</td></tr>
<tr><td style="width: 25%;"><a href="summary-ZlmFit-method.html">summary-method</a></td>
<td>Summarize model features from a 'ZlmFit' object</td></tr>
<tr><td style="width: 25%;"><a href="summary.thresholdSCRNACountMatrix.html">summary.thresholdSCRNACountMatrix</a></td>
<td>Summarize the effect of thresholding</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="thresholdSCRNACountMatrix.html">thresholdSCRNACountMatrix</a></td>
<td>Threshold a count matrix using an adaptive threshold.</td></tr>
</table>

<h2><a name="U">-- U --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="LMERlike-class.html">update-method</a></td>
<td>Wrapper for lmer/glmer</td></tr>
<tr><td style="width: 25%;"><a href="LMlike-class.html">update-method</a></td>
<td>Linear Model-like Class</td></tr>
</table>

<h2><a name="V">-- V --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="vbeta-dataset.html">vbeta</a></td>
<td>Vbeta Data Set</td></tr>
<tr><td style="width: 25%;"><a href="vbetaFA-dataset.html">vbetaFA</a></td>
<td>Vbeta Data Set, FluidigmAssay</td></tr>
<tr><td style="width: 25%;"><a href="GLMlike-class.html">vcov-method</a></td>
<td>Wrapper for regular glm/lm</td></tr>
<tr><td style="width: 25%;"><a href="LMERlike-class.html">vcov-method</a></td>
<td>Wrapper for lmer/glmer</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">vcov-method</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
</table>

<h2><a name="W">-- W --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="waldTest.html">waldTest</a></td>
<td>Run a Wald test</td></tr>
<tr><td style="width: 25%;"><a href="LMlike-class.html">waldTest-method</a></td>
<td>Linear Model-like Class</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">waldTest-method</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
<tr><td style="width: 25%;"><a href="waldTest-ZlmFit-matrix-method.html">waldTest-method</a></td>
<td>Wald test</td></tr>
</table>

<h2><a name="X">-- X --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="xform.html">xform</a></td>
<td>Make matrix of continuous expression values, orthogonal to discrete</td></tr>
</table>

<h2><a name="Z">-- Z --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="zlm.html">zlm</a></td>
<td>Zero-inflated regression for SingleCellAssay</td></tr>
<tr><td style="width: 25%;"><a href="zlm.html">zlm.SingleCellAssay</a></td>
<td>Zero-inflated regression for SingleCellAssay</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">ZlmFit</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
<tr><td style="width: 25%;"><a href="ZlmFit-class.html">ZlmFit-class</a></td>
<td>An S4 class to hold the output of a call to zlm</td></tr>
</table>

<h2><a name="misc">-- misc --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="summary-ZlmFit-method.html">"summary-ZlmFit",</a></td>
<td>Summarize model features from a 'ZlmFit' object</td></tr>
<tr><td style="width: 25%;"><a href="summary-ZlmFit-method.html">"summary.ZlmFit"</a></td>
<td>Summarize model features from a 'ZlmFit' object</td></tr>
<tr><td style="width: 25%;"><a href="summary-ZlmFit-method.html">"ZlmFit-summary",</a></td>
<td>Summarize model features from a 'ZlmFit' object</td></tr>
</table>
</div></body></html>
